aptos-move/move-examples/nfts_as_accounts/sources/hero.move

```diff
@@ -0,0 +1,287 @@
+/// This demonstrates how to make a composable NFT on Aptos. We take a Hero that can have optional
+/// weapons, armor, and a shield. Each of these can optionally have a gem. These can be added or
+/// removed and seamlessly treated as Tokens or their richer type.
+///
+/// todo(@davidiw):
+/// * complete armor and shield
+/// * add accessors to everything
+/// * make it possible to get a reference to the various ref types so that the items can remain
+///   as is without requiring decomposing
+/// * better error handling around if an item is already equipped
+/// * add the ability to remove weapons and gems
+module nfts_as_accounts::composable_example {
+    use std::option::{Self, Option};
+    use std::signer;
+    use std::string::{Self, String};
+
+    use nfts_as_accounts::token;
+
+    const ENOT_A_HERO: u64 = 1;
+    const ENOT_A_WEAPON: u64 = 2;
+    const ENOT_A_GEM: u64 = 3;
+
+    struct OnChainConfig has key {
+        collection: String,
+        mutability_config: token::MutabilityConfig,
+        royalty: token::Royalty,
+    }
+
+    struct Hero has key {
```

Comment:
What if instead of specifying shield / weapon I want to specify what they can equip left / right hand? Let's say that the options are shield and weapon. Is it possible to create an X or Y type?

Reply:
I think what you mean are Union types, which Move (and kinda most) languages doesn't support natively (TypeScript, Python 3 do). Interfaces aren't supported either.

I think what you could do is probably create some sort of wrapper struct like:

```
struct HandSlot has store {
  weapon: Option<Weapon>,
  shield: Option<Shield>,
}
```

and then
```
struct Hero has key {
  left_hand: HandSlot,
  right_hand: HandSlot
}
```

Reply:
This is quite ugly...but maybe doable...no idea how we could index this though.